### ISSUE TITLE
Integrate subprocess32

### ISSUE THREAD

**Issue by nirs** [naftaliharris/python2.8#51]
Already ported to 2.7, should be easy to integrate:
https://github.com/google/python-subprocess32

**Comment by wannaphongcom** [comment]
+1

**Comment by naftaliharris** [comment]
Hmm, shouldn't be too bad probably.

A quick look at the 3.5.2 version of subprocess shows that it requires time.monotonic, which I've been meaning to do as it's a dependency for some of the async-type stuff. But maybe makes sense to do 3.2 version of subprocess first as it's already done...

**Comment by kovidgoyal** [comment]
See https://github.com/kovidgoyal/calibre/blob/master/src/calibre/utils/monotonic.c for monotonic for python 2

**Comment by dan-stromberg** [comment]
I submitted a pull request to the subprocess32 project to make subprocess32 build on tauthon.  Someone e-mailed me back indicating it'd be better to merge subprocess32 into tauthon than to patch subprocess32 to work out of the box on tauthon.

**Comment by arizvisa** [comment]
@dan-stromberg, what's the PR# (or your branch) that you submitted to subprocess32?

**Comment by dan-stromberg** [comment]
https://github.com/google/python-subprocess32/pull/64

^_^

**Comment by arizvisa** [comment]
thx, sir.

I'm working on time.monotonic* right now. I'll look at how to include your PR afterwards.

**Comment by dan-stromberg** [comment]
Nice.

Thank you.

**Comment by arizvisa** [comment]
I just submitted Python3's `time.monotonic` functions as PR #122 (excluding `time.get_clock_info`). If something needs that lmk, and I can work on implementing that in the next couple days.

It needs testing for OSX as I don't have access to that platform. Actually it probably needs testing against whatever consumes it as I just wrote it up and submitted the PR a few minutes ago....

**Comment by arizvisa** [comment]
So the internal `_subprocess` and `_posixsubprocess` modules can probably be merged so that `fork_exec` is exposed on both platforms (and backwards compatibility can be maintained). Then anything that depends on these binary modules can just use whatever is necessary.

However, are there any other differences in the `subprocess` module between Python2 and Python3? If we simply replace `Lib/subprocess.py` with Python3's version, will that potentially break modules that depend on Python2's implementation? Is this something that we can just get away with, or would it just be safer in terms of compatibility to have it as a separate module similar to `subprocess32`?

**Comment by stefantalpalaru** [comment]
> If we simply replace Lib/subprocess.py with Python3's version, will that potentially break modules that depend on Python2's implementation? Is this something that we can just get away with

No, it isn't. Backwards compatibility is a very high priority for us - comes right after security.

**Comment by arizvisa** [comment]
I suppose it wasn't too clear, but according to https://docs.python.org/3/whatsnew/3.7.html, the differences seem to be that `subprocess.run` contains a `capture_output` keyword, an alias for `universal_newlines` as `text`, context manager for Popen, better handling of `KeyboardInterrupt` (by adding a delay), and `close_fds is now defaulted to true.

Other than `close_fds`, all of these things seem backwards-compatible. Is this all that's different between these interfaces?

Personally instead of using the `subprocess` interface, I prefer just having access to handles/fds and feeding i/o to coroutines that use `yield` as an expression, so I'm totally unfamiliar with the differences between both of the `subprocess` implementations.
